chore(docs): document how to set imagePullSecrets in baseJobTemplate

mitchnielsen opened the pull request:
## Summary

Documents how to set imagePullSecrets in the baseJobTemplate.

Will be expanded as part of https://linear.app/prefect/issue/PLA-1020/expand-documentation-on-how-to-use-basejobtemplate

Closes https://linear.app/prefect/issue/PLA-1029/document-how-to-set-image-pull-secrets-in-the-base-job-template

## Testing

I installed the `prefect-server` chart, and then deployed the `prefect-worker` chart with this config:

```yaml
worker:
  apiConfig: server
  serverApiConfig:
    apiUrl: http://prefect-server.prefect.svc.cluster.local:4200/api
  config:
    workPool: test-pool
    baseJobTemplate:
      configuration: |
        # ...
```

I then confirmed the configuration appeared as expected in the UI.

mitchnielsen committed:
Document how to set imagePullSecrets in baseJobTemplate

Documents how to set imagePullSecrets in the baseJobTemplate.

Will be expanded as part of https://linear.app/prefect/issue/PLA-1020/expand-documentation-on-how-to-use-basejobtemplate

Closes https://linear.app/prefect/issue/PLA-1029/document-how-to-set-image-pull-secrets-in-the-base-job-template

mitchnielsen committed:
Remove '.Values' from docs

'.Values' is how values are referenced from within Helm templates, not
how end users set it.

mitchnielsen commented on charts/prefect-worker/README.md:
```diff
@@ -241,17 +241,46 @@ prefect work-pool get-default-base-job-template --type kubernetes > base-job-tem
+       {
+         "job_configuration": {
+           "job_manifest": {
+             "spec": {
+               "template": {
+                 "spec": {
+                   "imagePullSecrets": "my-pull-secret"
```
```
$ k explain deployment.spec.template.spec.imagePullSecrets
GROUP:      apps
KIND:       Deployment
VERSION:    v1

FIELD: imagePullSecrets <[]LocalObjectReference>


DESCRIPTION:
    ImagePullSecrets is an optional list of references to secrets in the same
    namespace to use for pulling any of the images used by this PodSpec. If
    specified, these secrets will be passed to individual puller implementations
    for them to use. More info:
    https://kubernetes.io/docs/concepts/containers/images#specifying-imagepullsecrets-on-a-pod
    LocalObjectReference contains enough information to let you locate the
    referenced object inside the same namespace.

FIELDS:
  name  <string>
    Name of the referent. This field is effectively required, but due to
    backwards compatibility is allowed to be empty. Instances of this type with
    an empty value here are almost certainly wrong. More info:
    https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
```

mitchnielsen committed:
Fix JSON formatting of imagePullSecrets

mitchnielsen commented on charts/prefect-worker/README.md:
```diff
@@ -241,17 +241,50 @@ prefect work-pool get-default-base-job-template --type kubernetes > base-job-tem
+       {
+         "job_configuration": {
+           "job_manifest": {
+             "spec": {
+               "template": {
+                 "spec": {
+                   "imagePullSecrets": [
+                     {
+                       "name": "my-pull-secret"
+                     }
+                   ]
+                 }
+               }
```
```
       {
         "job_configuration": {
           "job_manifest": {
             "spec": {
               "template": {
                 "spec": {
                   "imagePullSecrets": [
                     {
                       "name": "my-pull-secret"
                     }
                   ]
                 }
               }
             }
           }
         }
       }
```

jamiezieziula commented on charts/prefect-worker/README.md:
```diff
@@ -241,17 +241,50 @@ prefect work-pool get-default-base-job-template --type kubernetes > base-job-tem
+worker:
+  config:
+    baseJobTemplate:
+      configuration: |
```
quick question -- if you provide this and nothing else, does this merge with the base base job template? or will it overwrite it to _just_ contain this information?

mitchnielsen replied:
Great question - I checked the [source code](https://github.com/PrefectHQ/prefect/blob/e38b549fa2ae4cd70c5a18bb0dc9cb34cf2ca1fa/src/prefect/cli/work_pool.py#L414-L415) and it looks like we don't merge it over the default, we just take whatever is given (using `prefect work-pool update`). Looks like this is a little different than providing overrides in `prefect.yaml`, for example.

As a test, I created a work pool with the default base job template added under `worker.config.baseJobTemplate`. Then I removed a field, in this case `restartPolicy`, and upgraded the chart. At that point, `restartPolicy` was no longer present. This confirms there's no merging with the base. I'll add a note 👍🏼

mitchnielsen committed:
Clarify that configuration is not merged with default

mitchnielsen committed:
Add back description of base job template

mitchnielsen commented on charts/prefect-worker/README.md:
````diff
@@ -233,25 +233,63 @@ pip install prefect-kubernetes
 ```
 
-2. Modify the base job template as needed
+2. Modify the base job template as needed. Keep in mind that modifications are not merged with the default template. The configuration
````
Fair point - I took some of that away when I realized there was a section above that gave higher-level context. I updated it there in a6bcffc.

jamiezieziula commented on charts/prefect-worker/README.md.gotmpl:
```diff
@@ -222,9 +222,13 @@ serviceAccount:
+`worker.config.baseJobTemplate.configuration`.
 
-1. Define the base job template in a local file. To get a formatted template, run the following command & store locally in `base-job-template.json`
+**Note**: if the target work pool (`config.workPool`) already exists, the base job template passed **will be ignored**.
```
hm good question. it _is_ true i think - but maybe with the addition of the init container it is no longer true for our helm charts.

mitchnielsen replied:
I bet that's it, since we're directly using `prefect work-pool update <json-file>` now. I'll remove that warning.

mitchnielsen committed:
Remove warning about BJT updates being ignored

Now that we have the initContainer that calls `prefect work-pool
update`, updates are indeed respected.
